File: leet-code/513.找树左下角的值.js
/*
 * @lc app=leetcode.cn id=513 lang=javascript
 *
 * [513] 找树左下角的值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findBottomLeftValue = function (root) {
  let maxDepth = -Infinity
  let res = 0
  const traversal = (node, depth) => {
    if (!node.left && !node.right) {
      // 叶子节点
      if (depth > maxDepth) {
        // 更新最深左孩子节点的值
        maxDepth = depth
        res = node.val
      }
    }

    if (node.left) {
      // 回溯
      depth++
      traversal(node.left, depth)
      depth--
    }

    if (node.right) {
      traversal(node.right, depth + 1)
    }
  }

  traversal(root, 1)
  return res
};
// @lc code=end

File: leet-code-rs/src/solutions/513_找树左下角的值.rs
/*
 * @lc app=leetcode.cn id=513 lang=rust
 *
 * [513] 找树左下角的值
 */
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[allow(unused)]
impl Solution {
    pub fn find_bottom_left_value(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn find_value(
            node: &Option<Rc<RefCell<TreeNode>>>,
            v: &mut i32,
            depth: &mut i32,
            cur_depth: i32,
        ) {
            if let Some(node) = node {
                let node = node.borrow();
                if &cur_depth > depth {
                    *v = node.val;
                    *depth = cur_depth;
                }

                let left = &node.left;
                find_value(left, v, depth, cur_depth + 1);

                let right = &node.right;
                find_value(right, v, depth, cur_depth + 1);
            }
        }

        let mut val = 0;
        let mut depth = 0;
        find_value(&root, &mut val, &mut depth, 1);

        return val;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_find_bottom_left_value() {
        //assert_eq!(Solution::find_bottom_left_value(vec![]),[]);
        assert!(true)
    }
}
